compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -1018,9 +1018,9 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
    */
   class MapToUnderlying extends TreeMap {
     override def transform(tree: Tree)(implicit ctx: Context): Tree = tree match {
-      case tree: Ident if tree.symbol.exists && !tree.symbol.owner.isClass && skipLocal(tree.symbol) =>
+      case tree: Ident if tree.symbol.exists && !tree.symbol.owner.isClass && !tree.symbol.isAnonymousFunction && skipLocal(tree.symbol) =>
```

Comment:
@liufengyun I added this other condition to fix a second bug with the same code

Reply:
The guard is getting quite long. Maybe extract it into a `def` with a comment explaining when a symbol is skipped

Reply:
It will be simpler to refactor this in the next PR I'm working on in which I already have to change the `skipLocal`

Reply:
#6180